src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsService`1.cs

```diff
@@ -2,50 +2,58 @@
 
 using System;
 using System.ComponentModel.Composition;
+using System.Threading.Tasks;
 
-using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+
+using IAsyncServiceProvider = Microsoft.VisualStudio.Shell.IAsyncServiceProvider;
 
 namespace Microsoft.VisualStudio.ProjectSystem.VS
 {
     /// <summary>
-    ///     Provides an implementation of <see cref="IVsService{T}"/> that calls into Visual Studio's <see cref="SVsServiceProvider"/>.
+    ///     Provides an implementation of <see cref="IVsService{T}"/> that calls into Visual Studio's <see cref="IAsyncServiceProvider"/>.
     /// </summary>
     [Export(typeof(IVsService<>))]
-    internal class VsService<T> : IVsService<T>
+    internal class VsService<T>
     {
+        private readonly AsyncLazy<T> _value;
+        private readonly IAsyncServiceProvider _serviceProvider;
         private readonly IProjectThreadingService _threadingService;
-        private readonly IServiceProvider _serviceProvider;
-        private readonly Type _serviceType;
 
         [ImportingConstructor]
-        public VsService([Import(typeof(SVsServiceProvider))]IServiceProvider serviceProvider, IProjectThreadingService threadingService)
-            : this(serviceProvider, threadingService, typeof(T))
-        {
-        }
-
-        protected VsService(IServiceProvider serviceProvider, IProjectThreadingService threadingService, Type serviceType)
+        public VsService([Import(typeof(SAsyncServiceProvider))]IAsyncServiceProvider serviceProvider, IProjectThreadingService threadingService)
         {
             Requires.NotNull(serviceProvider, nameof(serviceProvider));
             Requires.NotNull(threadingService, nameof(threadingService));
-            Requires.NotNull(serviceType, nameof(serviceType));
 
+            _value = new AsyncLazy<T>(GetServiceAsync, threadingService.JoinableTaskFactory);
             _serviceProvider = serviceProvider;
             _threadingService = threadingService;
-            _serviceType = serviceType;
         }
 
-        public T Value
+        public Task<T> GetValueAsync()
         {
-            get
-            {
-                _threadingService.VerifyOnUIThread();
+            return _value.GetValueAsync();
+        }
 
-                var service = (T)_serviceProvider.GetService(_serviceType);
+        protected virtual Type ServiceType
+        {
+            get { return typeof(T); }
+        }
+
+        private async Task<T> GetServiceAsync()
+        {
+            // If the service request requires a package load, GetServiceAsync will 
+            // happily do that on a background thread.
+            object iunknown = await _serviceProvider.GetServiceAsync(ServiceType)
+                                                    .ConfigureAwait(true);
```

Comment:
If we switch to the main thread no matter what why do we need to always return to the calling thread?  

Also, this means that all callers of `GetValueAsync` must know to `ConfigureAwait(true)` can you add a comment about that? I would prefer it if we returned a custom awaiter here so that the caller is always returned to their original thread regardless.

Reply:
Read this, https://github.com/Microsoft/vs-threading/blob/master/doc/cookbook_vs.md#should-i-await-a-task-with-configureawaitfalse. 

`ConfigureAwait(false)` == always switch to thread-pool thread. If we use `ConfigureAwait(false)`, and we ended up on UI thread in GetServiceAsync, we'll likely be forced on a thread pool thread, only to immediately switch back to the UI thread.

Look through history of teams, I want to change the project system's usage of ConfigureAwait to be always `ConfigureAwait(true)`.